# Introduction to Python for Finance - Part I
Created By: Cordell L. Tanny, CFA, FRM, FDP

Version: 2

Date of last revision: December 5, 2024

This notebook and all code are the intellectual property of Digital Hub Insights LLC. All rights reserved.

Suggested additional resource: https://www.w3schools.com/python

## Python Basics and Syntax

### The `print` statement

In [ ]:
# The print statement
print("I love quant finance")

I love quant finance


In [ ]:
# In Jupyter notebooks, you can also use display
display("machine learning or bust")

'machine learning or bust'

### Python variables

In [ ]:
# Variables
x = 5
y = 'Cordell'
print(x)
print(y)

5
Cordell


In [ ]:
# Casting
x = str(5)  # same as x = '5'
y = int(5)  # same as y = 5
z = float(5) # same as z = 5.0

print(x, y, z)

5 5 5.0


In [ ]:
# F-string formating in print statements
print(f'This is {x} as a string\n'
      f'This is {y} as an integer\n'
      f'This is {z} as a float')

This is 5 as a string
This is 5 as an integer
This is 5.0 as a float


### Basic python data types

In [ ]:
print(type(x))
print(type(y))
print(type(z))

<class 'str'>
<class 'int'>
<class 'float'>


In [ ]:
a = None
b = True

print(type(a))
print(type(b))

<class 'NoneType'>
<class 'bool'>


### Variable names

In [ ]:
camelCase = 'Cordell'
PascalCase = 'Cordell Tanny'
snake_case = 'Cordell Lewis Tanny'

print(camelCase + PascalCase + snake_case)

CordellCordell TannyCordell Lewis Tanny


In [ ]:
print(camelCase + " " + PascalCase + " " + snake_case)

Cordell Cordell Tanny Cordell Lewis Tanny


### Variable operations
Basic Math

In [ ]:
# Remember that z = 5.0 and y = 5
# Basic math operations
t = z + y
print(t)
print(type(t))

10.0
<class 'float'>


In [ ]:
# Division
z / y

1.0

In [ ]:
# Floor Division
x = 6
y = 4

print(x / y)  # normal division for a comparison
print(x // y)  # floor division


1.5
1


In [ ]:
# Think of floor division like this:
msft_share_price = 450
amount_to_invest = 100000
shares_to_buy = (amount_to_invest // msft_share_price)
print(f'Shares of MSFT to buy: {shares_to_buy}')

Shares of MSFT to buy: 222


In [ ]:
# Modulus - Returns the remainder
print(x % y)
print(amount_to_invest % msft_share_price)
cash_left_over = amount_to_invest - (shares_to_buy * msft_share_price)
print(cash_left_over)

2
100
100


In [ ]:
# Exponentiation
x ** y

1296

### Comparison operatos

In [ ]:
# Comparison operators
# 1) Greater than: returns boolean
z > y

True

In [ ]:
# 2) Greater than or equal to (or less than: "<=")
z >= y

True

In [ ]:
# Equality: Returns a Boolean
z == 6

False

In [ ]:
# Inequality
z != 6

True

### `and`, `or` and `not`

In [ ]:
# and statement
x == y and z != 6

False

In [ ]:
# more and statements
2 == 2.0 and 2 == 8 / 4

True

In [ ]:
# or statements
2 == 1.0 or 2 == 8 / 4

True

In [ ]:
# not statement: Reverses the result
not(2 == 2.0 and 2 == 8 / 4)

False

In [ ]:
not(2 == 1.0 or 2 == 8 / 4)

False

### More assignment operators

In [ ]:
x = 5
print(type(x))
x += 1  # same as x = x + 1
x

<class 'int'>


6

In [ ]:
x -= 1  # same as x = x -1
x

5

In [ ]:
x *= 3  # same as x = x * 3
x

15

In [ ]:
x /= 3  # same as x = x / 3
print(x)
print(type(x))

5.0
<class 'float'>


## Data Containers and Storage Types

### Lists
Lists are mutable:
You can change them, rearrange the order, add and remove items.

In [ ]:
list1 = ['Roger', 'Rafa', 'Novak']
list2 = list(('Coco', 'Aryna', 'Iga'))

print(list1)
print(type(list1))

['Roger', 'Rafa', 'Novak']
<class 'list'>


In [ ]:
# Determining how many entries
len(list1)

3

In [ ]:
# Accessing list elements - Introducing zero indexing
list1[0]

'Roger'

In [ ]:
list1[0] == 'Novak'

False

In [ ]:
list2[1] != 'David'

True

In [ ]:
# negative indexing
list1[-1]

'Novak'

In [ ]:
# negative indexing
list1[:2]

['Roger', 'Rafa']

In [ ]:
list1[-2:]

['Rafa', 'Novak']

In [ ]:
list1.index('Novak')

2

In [ ]:
# Determining how many items are in a list
list3 = ['Roger', 23, 5*3, ['Novak', 'Rafa']]
list3

['Roger', 23, 15, ['Novak', 'Rafa']]

In [ ]:
list3[3:]

[['Novak', 'Rafa']]

#### `is`, `is not`, `in`, `not in`

In [ ]:
'Rafa' in list1

True

In [ ]:
'Roger' in list2

False

In [ ]:
'Aryna' not in list1

True

In [ ]:
list1[0] is 'Roger'

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<ipython-input-42-22e79ceade7d>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  list1[0] is 'Roger'


True

In [ ]:
list2[0] is not 'Roger'

<>:1: SyntaxWarning: "is not" with a literal. Did you mean "!="?
<>:1: SyntaxWarning: "is not" with a literal. Did you mean "!="?
<ipython-input-43-fed59e2a771f>:1: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  list2[0] is not 'Roger'


True

#### Adding, changing and removing list items

In [ ]:
# Remember:
list1[1:2]

['Rafa']

In [ ]:
# To add to the end of  the list
list1.append('Andy')
list1

['Roger', 'Rafa', 'Novak', 'Andy']

In [ ]:
# To add something at a specific location
list1.insert(0, 'Jannik')
list1

['Jannik', 'Roger', 'Rafa', 'Novak', 'Andy']

In [ ]:
list2.insert(2, 'Elena')
list2

['Coco', 'Aryna', 'Elena', 'Iga']

In [ ]:
# Removing items
list2.remove('Iga')
list2

['Coco', 'Aryna', 'Elena']

In [ ]:
# Remove based on index
list1.pop(3)
list1

['Jannik', 'Roger', 'Rafa', 'Andy']

In [ ]:
popped = list1.pop(0)
popped

'Jannik'

#### Sorting list items


In [ ]:
# Sorting
list1.sort()
list1

['Andy', 'Rafa', 'Roger']

In [ ]:
# Sorting descending
list1.sort(reverse=True)
list1

['Roger', 'Rafa', 'Andy']

#### Copying and joining

In [ ]:
list4 = list2.copy()
list4

['Coco', 'Aryna', 'Elena']

In [ ]:
# 1) Use +
list5 = list1 + list2
list5

['Roger', 'Rafa', 'Andy', 'Coco', 'Aryna', 'Elena']

In [ ]:
# 2) Use extend
list4 = list1.copy()
list4.extend(list2)
print(list4)

['Roger', 'Rafa', 'Andy', 'Coco', 'Aryna', 'Elena']


### Sets
- Sets are immutable
- Cannot be accessed with indexers
- Sets are unordered

Sets are great to use when you want to remove duplicates from a list and focus on unique values only.

In [ ]:
my_set = {'MSFT', 'AAPL', 'META'}
print(type(my_set))

<class 'set'>


In [ ]:
my_set2 = {'MSFT', 'AAPL', 'META', 'MSFT'}

print(my_set2)

{'MSFT', 'META', 'AAPL'}


In [ ]:
list_with_duplicates = ['MSFT', 'AAPL', 'MSFT', 'NFLX']
list_with_duplicates

['MSFT', 'AAPL', 'MSFT', 'NFLX']

In [ ]:
set_3 = set(list_with_duplicates)
set_3

{'AAPL', 'MSFT', 'NFLX'}

### Tuples
Tuples are used to store multiple items in the same variable
- They are ordered.
- Immutable.
- Allow duplicates.
- Must have a comma somewhere!
- Support indexing and functions in the same way that you access lists.

In [ ]:
tuple_1 = ('MSFT', )
print(type(tuple_1))

<class 'tuple'>


In [ ]:
tuple_2 = ('MSFT')
print(type(tuple_2))

<class 'str'>


In [ ]:
tuple_3 = (True, True, False, 34)
tuple_3[1]

True

In [ ]:
tuple_4 = tuple(('MSFT', 'META', 'AMZN', 'MSFT'))
print(tuple_4)

('MSFT', 'META', 'AMZN', 'MSFT')


In [ ]:
tuple_4[1:3]

('META', 'AMZN')

In [ ]:
'MSFT' in tuple_4

True

### Dictionaries
A dictionary in Python is like a real-world dictionary:

it pairs a "key" with a "value." For example, in a dictionary of stocks, the stock ticker (like AAPL for Apple) can be the key, and the company name (like "Apple Inc.") can be the value.

Think of a dictionary as a container that stores information in pairs, organized like this:

Key: Value

Example:

'MSFT': 'Microsoft Corporation'

Why Use Dictionaries?
Dictionaries are powerful because:

You can quickly look up data by its key.
They are easy to update and manage.
They help you organize related information (like stock tickers and names).

In [ ]:
# Creating a dictionary
stocks = {
    "AAPL": "Apple Inc.",
    "MSFT": "Microsoft Corporation",
    "GOOGL": "Alphabet Inc.",
    "AMZN": "Amazon.com Inc."
}

print(stocks)


{'AAPL': 'Apple Inc.', 'MSFT': 'Microsoft Corporation', 'GOOGL': 'Alphabet Inc.', 'AMZN': 'Amazon.com Inc.'}


#### Accessing dictionary values from keys

In [ ]:
# Accessing a value by its key
print(stocks["AAPL"])
print(stocks["MSFT"])

Apple Inc.
Microsoft Corporation


#### Adding new keys or updating existing ones

In [ ]:
# Adding a new stock
stocks["TSLA"] = "Tesla Inc."

# Updating an existing stock
stocks["GOOGL"] = "Google LLC"

print(stocks)


{'AAPL': 'Apple Inc.', 'MSFT': 'Microsoft Corporation', 'GOOGL': 'Google LLC', 'AMZN': 'Amazon.com Inc.', 'TSLA': 'Tesla Inc.'}


In [ ]:
# Removing a stock
del stocks["AMZN"]
print(stocks)


{'AAPL': 'Apple Inc.', 'MSFT': 'Microsoft Corporation', 'GOOGL': 'Google LLC', 'TSLA': 'Tesla Inc.'}


In [ ]:
# Check if a key exists
print("AAPL" in stocks)
print("NFLX" in stocks)


True
False


#### Dictionaries of Lists

Dictionaries of lists are helpful when you:

1. Need to store multiple pieces of data for each key.
2. Want to keep related information organized in one place.
3. Need quick access to all or specific data for each key.


In [ ]:
# Dictionary of lists: Stock tickers and their recent prices
stock_prices = {
    "AAPL": [150.25, 151.30, 149.80],  # Prices for Apple Inc.
    "MSFT": [299.50, 300.00, 298.20],  # Prices for Microsoft
    "GOOGL": [2800.25, 2825.00, 2790.50]  # Prices for Alphabet Inc.
}

# Accessing the list of prices for a stock
print(stock_prices["AAPL"])  # Output: [150.25, 151.30, 149.80]

# Accessing a specific price
print(stock_prices["AAPL"][0])  # Output: 150.25


[150.25, 151.3, 149.8]
150.25


In [ ]:
# Dictionary of lists: Stock tickers and their sectors
stock_sectors = {
    "AAPL": ["Technology", "Consumer Electronics"],
    "MSFT": ["Technology", "Software"],
    "XOM": ["Energy", "Oil & Gas"]
}

# Accessing the list of sectors for a stock
print(stock_sectors["MSFT"])  # Output: ['Technology', 'Software']

# Accessing the main sector
print(stock_sectors["XOM"][0])  # Output: Energy


['Technology', 'Software']
Energy


In [ ]:
# Dictionary of lists: Stock tickers and their performance metrics
stock_performance = {
    "AAPL": [10.5, 3.2, 15.7],  # % returns: Year-to-date, 1-year, 5-year
    "MSFT": [8.0, 2.5, 20.1],
    "GOOGL": [12.3, 4.1, 18.9]
}

# Accessing performance data for a stock
print(stock_performance["GOOGL"])

# Accessing a specific performance metric (e.g., 1-year return)
print(stock_performance["MSFT"][1])


[12.3, 4.1, 18.9]
2.5


#### Nested Dictionaries
A nested dictionary is a dictionary where each key has another dictionary as its value. This is useful when you want to store multiple pieces of information for each key in a structured way.

Nested dictionaries are helpful when:

1. You need to organize data hierarchically.
2. Each key needs to store multiple categories of data (like prices, sectors, and performance metrics).
3. You want quick access to specific data points using clear keys.

In [ ]:
# Nested dictionary: Stock tickers and their recent prices
stock_prices = {
    "AAPL": {"Day1": 150.25, "Day2": 151.30, "Day3": 149.80},
    "MSFT": {"Day1": 299.50, "Day2": 300.00, "Day3": 298.20},
    "GOOGL": {"Day1": 2800.25, "Day2": 2825.00, "Day3": 2790.50}
}

# Accessing the dictionary for a stock
print(stock_prices["AAPL"])

# Accessing a specific day's price
print(stock_prices["AAPL"]["Day1"])


{'Day1': 150.25, 'Day2': 151.3, 'Day3': 149.8}
150.25


In [ ]:
# Nested dictionary: Stock tickers and their sectors
stock_sectors = {
    "AAPL": {"Sector": "Technology", "Sub-sector": "Consumer Electronics"},
    "MSFT": {"Sector": "Technology", "Sub-sector": "Software"},
    "XOM": {"Sector": "Energy", "Sub-sector": "Oil & Gas"}
}

# Accessing the dictionary for a stock
print(stock_sectors["MSFT"])

# Accessing the main sector
print(stock_sectors["XOM"]["Sector"])


{'Sector': 'Technology', 'Sub-sector': 'Software'}
Energy


In [ ]:
# Nested dictionary: Stock tickers and their performance metrics
stock_performance = {
    "AAPL": {"YTD": 10.5, "1-Year": 3.2, "5-Year": 15.7},
    "MSFT": {"YTD": 8.0, "1-Year": 2.5, "5-Year": 20.1},
    "GOOGL": {"YTD": 12.3, "1-Year": 4.1, "5-Year": 18.9}
}

# Accessing the dictionary for a stock
print(stock_performance["GOOGL"])  # Output: {'YTD': 12.3, '1-Year': 4.1, '5-Year': 18.9}

# Accessing a specific performance metric (e.g., 1-year return)
print(stock_performance["MSFT"]["1-Year"])  # Output: 2.5


{'YTD': 12.3, '1-Year': 4.1, '5-Year': 18.9}
2.5


In [ ]:
# Comprehensive nested dictionary: Stock data
stocks = {
    "AAPL": {
        "Prices": {"Day1": 150.25, "Day2": 151.30, "Day3": 149.80},
        "Sector": {"Main": "Technology", "Sub-sector": "Consumer Electronics"},
        "Performance": {"YTD": 10.5, "1-Year": 3.2, "5-Year": 15.7}
    },
    "MSFT": {
        "Prices": {"Day1": 299.50, "Day2": 300.00, "Day3": 298.20},
        "Sector": {"Main": "Technology", "Sub-sector": "Software"},
        "Performance": {"YTD": 8.0, "1-Year": 2.5, "5-Year": 20.1}
    },
    "GOOGL": {
        "Prices": {"Day1": 2800.25, "Day2": 2825.00, "Day3": 2790.50},
        "Sector": {"Main": "Technology", "Sub-sector": "Internet Services"},
        "Performance": {"YTD": 12.3, "1-Year": 4.1, "5-Year": 18.9}
    }
}

# Accessing all data for AAPL
print(stocks["AAPL"])

# Accessing a specific price for AAPL
print(stocks["AAPL"]["Prices"]["Day1"])  # Output: 150.25

# Accessing the sector for MSFT
print(stocks["MSFT"]["Sector"]["Main"])  # Output: Technology

# Accessing the 1-year return for GOOGL
print(stocks["GOOGL"]["Performance"]["1-Year"])  # Output: 4.1


{'Prices': {'Day1': 150.25, 'Day2': 151.3, 'Day3': 149.8}, 'Sector': {'Main': 'Technology', 'Sub-sector': 'Consumer Electronics'}, 'Performance': {'YTD': 10.5, '1-Year': 3.2, '5-Year': 15.7}}
150.25
Technology
4.1


In [ ]:
# Lastly, you can always copy of dictionary so that
# you can mess around with a different version
stock_dict_temp = stocks.copy()
stock_dict_temp

{'AAPL': {'Prices': {'Day1': 150.25, 'Day2': 151.3, 'Day3': 149.8},
  'Sector': {'Main': 'Technology', 'Sub-sector': 'Consumer Electronics'},
  'Performance': {'YTD': 10.5, '1-Year': 3.2, '5-Year': 15.7}},
 'MSFT': {'Prices': {'Day1': 299.5, 'Day2': 300.0, 'Day3': 298.2},
  'Sector': {'Main': 'Technology', 'Sub-sector': 'Software'},
  'Performance': {'YTD': 8.0, '1-Year': 2.5, '5-Year': 20.1}},
 'GOOGL': {'Prices': {'Day1': 2800.25, 'Day2': 2825.0, 'Day3': 2790.5},
  'Sector': {'Main': 'Technology', 'Sub-sector': 'Internet Services'},
  'Performance': {'YTD': 12.3, '1-Year': 4.1, '5-Year': 18.9}}}

#### Casting to Create a Dictionary
You can use the dict() function to convert data into a dictionary. Let’s use stock tickers and their company names as an example.

Casting is useful when:

1. You already have data in lists, tuples, or other formats.
2. You want to transform the data into a structured format for easier access and management.
3. You need a quick way to build a dictionary without manually defining it.

In [ ]:
# List of tuples: (Ticker, Company Name)
stock_data = [("AAPL", "Apple Inc."), ("MSFT", "Microsoft Corporation"), ("GOOGL", "Alphabet Inc.")]

# Casting to dictionary
stocks = dict(stock_data)

print(stocks)


{'AAPL': 'Apple Inc.', 'MSFT': 'Microsoft Corporation', 'GOOGL': 'Alphabet Inc.'}


In [ ]:
# Two separate lists
tickers = ["AAPL", "MSFT", "GOOGL"]
company_names = ["Apple Inc.", "Microsoft Corporation", "Alphabet Inc."]

# Pairing and casting to dictionary
stocks = dict(zip(tickers, company_names))

print(stocks)


{'AAPL': 'Apple Inc.', 'MSFT': 'Microsoft Corporation', 'GOOGL': 'Alphabet Inc.'}


What does `zip` do?

The `zip()` function in Python takes two or more iterables (like lists, tuples, etc.) and combines them into pairs (or tuples), element by element. It creates a zip object, which is an iterator of these paired elements.

In [ ]:
list1 = [1, 2, 3]
list2 = ['a', 'b', 'c']

zipped = zip(list1, list2)
print(list(zipped))


[(1, 'a'), (2, 'b'), (3, 'c')]


**An important note about zip()**

- zip() produces tuples because it’s designed to pair elements generically.

- To turn the pairs into a dictionary, you need to explicitly use dict(zip(...)), as dictionaries require key-value pairs. Without the dict() function, zip() alone has no knowledge of "keys" or "values."

---



## Summary of Lists, Sets, Tuples, Dictionaries

Here is a brief comparison of all of these data types and when you should use them:

1. **Lists**
- Key Features:

 - Mutable (can be changed after creation).
 - Ordered (elements have a specific sequence).
 - Allows duplicate elements.

- Best for:

 - Storing ordered data where duplicates are allowed.
 - Scenarios requiring frequent additions, deletions, or updates to elements.
 - Iterative tasks, such as looping through a collection of items.

- Example Use Cases:
 - Daily stock prices in a portfolio.
 - A to-do list of tasks.

2. **Sets**
- Key Features:

 - Mutable (but all elements must be immutable).
 - Unordered (elements are not stored in a specific sequence).
 - No duplicate elements allowed.

- Best for:

 - Ensuring uniqueness in a collection.
 - Fast membership tests (element in set).
 - Mathematical operations like unions, intersections, and differences.

- Example Use Cases:

 - A collection of unique tickers in a portfolio.
 - Comparing the tickers in two stock indices to find common or distinct ones.

3. **Tuples**
- Key Features:

 - Immutable (cannot be changed after creation).
 - Ordered (elements have a specific sequence).
 - Allows duplicate elements.

- Best for:
 - Data that should not change during the program’s lifecycle.
 - Fixed collections of related elements that belong together conceptually.

- Example Use Cases:
 - Latitude and longitude coordinates for a stock exchange.
 - Fixed settings like (Ticker, Exchange, Sector).

4. **Dictionaries**
- Key Features:

 - Mutable (can be changed after creation).
 - Unordered (insertion order is preserved as of Python 3.7+).
 - Stores data as key-value pairs.
 - Keys must be unique, values can be duplicated.

- Best for:

 - Mapping relationships between items.
 - Quick lookups by a unique key.
 - Complex structures where each item has multiple attributes.

- Example Use Cases:

 - A mapping of stock tickers to company names (e.g., {'AAPL': 'Apple Inc.'}).
 - Financial metrics for a stock (e.g., {'AAPL': {'Price': 150, 'PE Ratio': 25}}).
